daemon/containerd/image_list.go

```diff
@@ -337,6 +346,111 @@ func (i *ImageService) setupFilters(ctx context.Context, imageFilters filters.Ar
 	}, nil
 }
 
+// setupLabelFilter parses filter args for "label" and "label!" and returns a
+// filter func which will check if any image config from the given image has
+// labels that match given predicates.
+func setupLabelFilter(store content.Store, fltrs filters.Args) (func(image images.Image) bool, error) {
+	type labelCheck struct {
+		key        string
+		value      string
+		onlyExists bool
+		negate     bool
+	}
+
+	var checks []labelCheck
+	for _, fltrName := range []string{"label", "label!"} {
+		for _, l := range fltrs.Get(fltrName) {
+			k, v, found := strings.Cut(l, "=")
+			err := labels.Validate(k, v)
+			if err != nil {
+				return nil, err
+			}
+
+			negate := strings.HasSuffix(fltrName, "!")
+
+			// If filter value is key!=value then flip the above.
+			if strings.HasSuffix(k, "!") {
+				k = strings.TrimSuffix(k, "!")
+				negate = !negate
+			}
```

Comment:
This is confusing me; are we checking the same thing twice here? (Is this correct?)

Reply:
No, consider this: 
`label!=some.image.label=asdf`

fltrName=`label!`
k=`some.image.label`
v=`asdf`

`negate := strings.HasSuffix(fltrName, "!")` checks if the filter name implies negation.
`if strings.HasSuffix(k, "!") {` checks if the "operator" is `!=`


This is a bit awkward, because now we have both `label!` as a filter (to be backwards compatible with non-c8d Docker) and `!=` as a operator filter expression (to make it consistent with containerd filters, and we probably might want to switch completely to them in future).
So something like this will also work: `label!=some.image.label!=asdf` (but would be equal to `label=some.image.label=asdf`).

Reply:
> This is a bit awkward, because now we have both label! as a filter (to be backwards compatible with non-c8d Docker) and != as a operator filter expression

Yeah the `label!` was one giant hack, but I wondered if we weren't already handling that with the first step (if we see the filter "name" is `label!` we could trim that, and set the operator to `!=`.

(I'm rambling; I need to have another look at the code)


> So something like this will also work: label!=some.image.label!=asdf (but would be equal to label=some.image.label=asdf).

Did that work before? 🤯 I wonder if we ever designed that, or if it's (like many things with our filtering) just "happened to be there".

Reply:
> Did that work before? 🤯

Yep! (it was available only for `image prune` though, not `image list`)

```bash
$ docker info -f '{{.Driver}}'
overlay2

$ docker image prune -a --filter 'label!=org.opencontainers.image.version!=22.04'
WARNING! This will remove all images without at least one container associated to them.
Are you sure you want to continue? [y/N] y
Deleted Images:
untagged: ubuntu:latest
untagged: ubuntu@sha256:67211c14fa74f070d27cc59d69a7fa9aeff8e28ea118ef3babc295a0428a6d21
deleted: sha256:bab8ce5c00ca3ef91e0d3eb4c6e6d6ec7cffa9574c447fd8d54a8d96e7c1c80e
deleted: sha256:874b048c963ab55b06939c39d59303fb975d323822a4ea48a02ac8dc635ea371

```